File: examples/java/core/shared/AppStructuralContracts.java
package core.shared;

import java.util.List;

public final class AppStructuralContracts {
  private AppStructuralContracts() {}

  public record AppError(String code, String message, Object details) {}

  public static class AppCaseError extends RuntimeException {
    private final String code;
    private final Object details;

    public AppCaseError(String code, String message) {
      this(code, message, null);
    }

    public AppCaseError(String code, String message, Object details) {
      super(message);
      this.code = code;
      this.details = details;
    }

    public String code() {
      return code;
    }

    public Object details() {
      return details;
    }

    public AppError toAppError() {
      return new AppError(code, getMessage(), details);
    }
  }

  public static AppCaseError toAppCaseError(
      AppError error,
      String fallbackMessage,
      String fallbackCode,
      Object fallbackDetails
  ) {
    if (error != null) {
      return new AppCaseError(error.code(), error.message(), error.details());
    }

    return new AppCaseError(fallbackCode, fallbackMessage, fallbackDetails);
  }

  public static AppCaseError toAppCaseError(AppError error, String fallbackMessage) {
    return toAppCaseError(error, fallbackMessage, "INTERNAL", null);
  }

  public static class AppResult<T> {
    public boolean success;
    public T data;
    public AppError error;

    public static <T> AppResult<T> success(T data) {
      AppResult<T> result = new AppResult<>();
      result.success = true;
      result.data = data;
      return result;
    }

    public static <T> AppResult<T> failure(AppError error) {
      AppResult<T> result = new AppResult<>();
      result.success = false;
      result.error = error;
      return result;
    }
  }

  public record StreamFailureEnvelope<TEvent>(
      String caseName,
      String surface,
      TEvent originalEvent,
      FailureError lastError,
      int attempts,
      String firstAttemptAt,
      String lastAttemptAt,
      String correlationId
  ) {
    public record FailureError(String message, String code, String stack) {}
  }

  public static class AppPaginationParams {
    public Integer page;
    public Integer limit;
    public String cursor;
  }

  public static class AppPaginatedResult<T> {
    public List<T> items;
    public Integer total;
    public Integer page;
    public Integer limit;
    public String cursor;
    public Boolean hasMore;
  }
}

File: examples/go/core/shared/app_structural_contracts.go
package shared

import "errors"

type AppError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
	Details any    `json:"details,omitempty"`
}

type AppCaseError struct {
	Code    string
	Message string
	Details any
}

func (e *AppCaseError) Error() string {
	return e.Message
}

func (e *AppCaseError) ToAppError() AppError {
	return AppError{
		Code:    e.Code,
		Message: e.Message,
		Details: e.Details,
	}
}

func ToAppCaseError(err error, fallbackMessage string, fallbackCode string, fallbackDetails any) *AppCaseError {
	if err == nil {
		return &AppCaseError{
			Code:    fallbackCode,
			Message: fallbackMessage,
			Details: fallbackDetails,
		}
	}

	var caseErr *AppCaseError
	if errors.As(err, &caseErr) {
		return caseErr
	}

	return &AppCaseError{
		Code:    fallbackCode,
		Message: fallbackMessage,
		Details: fallbackDetails,
	}
}

func ToAppError(err error, fallbackCode string) AppError {
	var caseErr *AppCaseError
	if errors.As(err, &caseErr) {
		return caseErr.ToAppError()
	}

	return AppError{
		Code:    fallbackCode,
		Message: err.Error(),
	}
}

type AppResult[T any] struct {
	Success bool      `json:"success"`
	Data    *T        `json:"data,omitempty"`
	Error   *AppError `json:"error,omitempty"`
}

type StreamFailureEnvelope struct {
	CaseName      string `json:"caseName"`
	Surface       string `json:"surface"`
	OriginalEvent any    `json:"originalEvent"`
	LastError     struct {
		Message string `json:"message"`
		Code    string `json:"code,omitempty"`
		Stack   string `json:"stack,omitempty"`
	} `json:"lastError"`
	Attempts       int    `json:"attempts"`
	FirstAttemptAt string `json:"firstAttemptAt"`
	LastAttemptAt  string `json:"lastAttemptAt"`
	CorrelationID  string `json:"correlationId"`
}

type AppPaginationParams struct {
	Page   int    `json:"page,omitempty"`
	Limit  int    `json:"limit,omitempty"`
	Cursor string `json:"cursor,omitempty"`
}

type AppPaginatedResult[T any] struct {
	Items      []T    `json:"items"`
	TotalItems int    `json:"totalItems,omitempty"`
	NextCursor string `json:"nextCursor,omitempty"`
	Page       int    `json:"page,omitempty"`
	Limit      int    `json:"limit,omitempty"`
}
